### DISCUSSION TITLE
Help combining two scripts !!

### DISCUSSION THREAD

**Discussion by furycd001** [ImageMagick/ImageMagick#7749] [Help] [answered]
### ImageMagick version

7.1.1-39

### Operating system, version and so on

Arch Linux (fully up-to-date)

### Description

HI..

I'm working on a script that processes images using two separate ImageMagick-based bash scripts. These scripts work independently but, when combined, I encounter a problem where the output images end up partially or completely in grayscale. I’d like the combined script to fully process each image before moving on to the next one, but I haven’t been able to get that working either. As a beginner with ImageMagick, I’d really appreciate any advice on how to effectively combine these two scripts while avoiding the grayscale issues I’ve been encountering.

### Script 1: Creates a Tiled Background
- This script generates a tiled background from the top portion of each input image.
- It extracts a narrow strip from the top, alternates between a normal and flipped version, and tiles these to fill the target dimensions (1440x3216).
- The final tiled background has a blur applied across it, saved in an output folder.

### Script 2: Composites the Original Image onto the Tiled Background
- This script overlays the original image onto the background created by Script 1, aligning it at the bottom of the canvas.
- The resulting composite is saved in the same output folder.

**Current Issues When Combining the Scripts:**
1. **Grayscale Output**: When combined, either parts of the image or the entire output come out in grayscale. I’ve tried explicitly setting the color space to `sRGB` in several places, but the problem persists.
2. **Sequential Processing:** Ideally, I’d like the combined script to fully process each image from start to finish (creating the tiled background and then applying the overlay) before moving on to the next image. In their current form, the scripts require that all tiled backgrounds be completed first before the overlay process in the second script can even begin. This setup slows down processing, and I’d like help making it way more efficient.

**Could anyone help with combining these scripts in a way that:**
1. Retains full color in the output image.
2. Completes the entire process for one image before moving to the next.

----

I’ve attached the two scripts in a `.zip` file to this post & also uploaded them to termbin..

+ **[imager.zip](https://github.com/user-attachments/files/17630982/imager.zip)**
+ **[Script 1](https://termbin.com/0c1q)**
+ **[Script 2](https://termbin.com/kqec)**

On my own system, I have these saved in `~/.local/bin/`  

Thank you so much in advance!! If you've read this far, I hope you have a wonderful and totally awesome day!

**Comment by snibgo** [comment]
In `imagerone` you have:
```
        # Ensure the extracted tile remains in full color
        magick "$temp_tile" -set colorspace sRGB "$temp_tile"
```
That does nothing. If the input is grayscale so has only one channel, then after IM has read the image, `-set colorspace sRGB` will expand the image to 3 channels. But then IM will check whether the image is grayscale (all channels are equal), and will write to the file accordingly, so the file will have only one channel.

The grayscale problem occurs when an operation such as `-composite` has two inputs: the first is grayscale, the second is colour. Then the output of the operation will be grayscale. The fix is to `-set colorspace sRGB` after the first input, or after both inputs, and before the operation. You do this, so I don't see any problem.

If you do find unexpected grayscale images, you should isolate the command that creates it. Then, if you can't fix the problem, post that command and its input files here.

**Reply by furycd001** [reply, marked as answer]
Thank you very much for the detailed explanation. Much appreciated., I literally just managed to fix my grayscale problem with the code below....

```bash
#!/usr/bin/env bash

# Define target dimensions
target_width=1440
target_height=3216

# Create an output directory if it doesn't exist
output_folder="./output"
mkdir -p "$output_folder"

# Path to sRGB color profile
sRGB_profile="/usr/share/color/icc/colord/sRGB.icc"

# Process each image in the current directory
for img in *.jpg; do
    if [[ -f "$img" ]]; then
        # Get the base name of the image without the extension
        base_name=$(basename "$img" .jpg)
        
        # Define paths for temporary and final images
        temp_tile="$output_folder/${base_name}_tile_temp.png"
        temp_tile_flipped="$output_folder/${base_name}_tile_temp_flipped.png"
        tiled_output="$output_folder/${base_name}_tiled_output.png"
        final_output="$output_folder/${base_name}_final.png"

        # Extract the top portion of the image in color mode
        magick "$img" -colorspace sRGB -gravity North -crop "${target_width}x22+0+0" +repage "$temp_tile"

        # Ensure the extracted tile remains in full color
        magick "$temp_tile" -colorspace sRGB -type TrueColor "$temp_tile"

        # Create a flipped version of the tile
        magick "$temp_tile" -flip -type TrueColor "$temp_tile_flipped"

        # Initialize an empty RGB canvas with the target dimensions
        magick -size "${target_width}x${target_height}" canvas:none -colorspace sRGB -type TrueColor "$tiled_output"

        # Calculate the number of tiles needed
        num_tiles=$((target_height / 22 + 1))

        # Alternate placing the tiles in the target image
        for ((i = 0; i < num_tiles; i++)); do
            offset_y=$((i * 22))
            tile_to_use="$temp_tile"
            [[ $((i % 2)) -ne 0 ]] && tile_to_use="$temp_tile_flipped"

            # Composite the tile, ensuring color is maintained
            magick "$tiled_output" "$tile_to_use" -colorspace sRGB -type TrueColor -geometry +0+"$offset_y" -composite "$tiled_output"
        done

        # Apply 22% blur to the entire tiled output
        magick "$tiled_output" -blur 0x22 -colorspace sRGB -type TrueColor "$tiled_output"

        # Create the final output with the original image at the bottom
        magick "$tiled_output" -colorspace sRGB \
            "$img" -gravity South -composite -type TrueColor "$final_output"
        
        echo "Final image created: $final_output"

        # Clean up temporary images
        rm "$temp_tile" "$temp_tile_flipped" "$tiled_output"
    fi
done

echo "All final images saved in the output folder."
```
